### PR TITLE
make `CertificateParams::signed_by` accept `impl PublicKeyData` rather than `KeyPair` for the key being signed

### PR DESCRIPTION
**Pull request by kwantam**
`CertificateParams::signed_by` does not need the secret corresponding to the public key being certified (though of course it needs the secret for the CA key signing the certificate!). More discussion about this in #282.

This PR:
- changes the first arg of `signed_by` to `impl PublicKeyData` (which maintains compatibility because `KeyPair` implements this trait)
- adds a new type, `SubjectPublicKey`, that can be parsed from a PEM or DER output by this crate (or, e.g., by openssl) and also impls `PublicKeyData.` This type is only available when the `x509-parser` feature is enabled (because that's what is required to parse).

I added a quick test that the `SubjectPublicKey` type's `PublicKeyData` impl produces the same output as `KeyPair` would for the same key, at least for the key types that can be generated using the crate. (So this means, no test right now for RSA keys, since those seemed to give an error when I tried to generate a keypair.)

### CODE REVIEW

File: rcgen/Cargo.toml [hunk 1]
```diff
@@ -36,7 +36,7 @@ x509-parser = { workspace = true, features = ["verify"], optional = true }
 zeroize = { version = "1.2", optional = true }
 
 [features]
-default = ["crypto", "pem", "ring"]
+default = ["crypto", "pem", "ring", "x509-parser"]
```
Comment (djc) [inline rcgen/Cargo.toml:39]: We'll want to revert this.

File: rcgen/src/certificate.rs [hunk 2]
```diff
@@ -147,9 +147,9 @@ impl CertificateParams {
 	///
 	/// The returned [`Certificate`] may be serialized using [`Certificate::der`] and
 	/// [`Certificate::pem`].
-	pub fn signed_by(
+	pub fn signed_by<K: PublicKeyData>(
 		self,
-		key_pair: &KeyPair,
+		key_pair: &K,
```
Comment (djc) [inline rcgen/src/certificate.rs:152]: Nit: I'd prefer to spell this as `&impl PublicKeyData`.

File: rcgen/src/key_pair.rs [hunk 3]
```diff
@@ -57,6 +57,90 @@ impl fmt::Debug for KeyPairKind {
 	}
 }
 
+/// A public key
+#[cfg(feature = "x509-parser")]
```
Comment (djc) [inline rcgen/src/key_pair.rs:61]: There's no need to guard the type directly.

File: rcgen/src/key_pair.rs [hunk 4]
```diff
@@ -57,6 +57,90 @@ impl fmt::Debug for KeyPairKind {
...
+impl SubjectPublicKey {
+	/// Create a `SubjectPublicKey` value from a PEM-encoded SubjectPublicKeyInfo string
+	pub fn from_pem(pem_str: &str) -> Result<Self, Error> {
+		let spki_der = pem::parse(pem_str)._err()?.into_contents();
+		Self::from_der(&spki_der)
+	}
+
+	/// Create a `SubjectPublicKey` value from DER-encoded SubjectPublicKeyInfo bytes
+	pub fn from_der(spki_der: &[u8]) -> Result<Self, Error> {
```
Comment (djc) [inline rcgen/src/key_pair.rs:77]: We should guard this constructor.

File: rcgen/src/key_pair.rs [hunk 5]
```diff
@@ -57,6 +57,90 @@ impl fmt::Debug for KeyPairKind {
...
+#[derive(Debug)]
+pub struct SubjectPublicKey {
+	pub(crate) alg: &'static SignatureAlgorithm,
+	pub(crate) subject_public_key: Vec<u8>,
+}
+
+#[cfg(feature = "x509-parser")]
+impl SubjectPublicKey {
+	/// Create a `SubjectPublicKey` value from a PEM-encoded SubjectPublicKeyInfo string
+	pub fn from_pem(pem_str: &str) -> Result<Self, Error> {
```
Comment (djc) [inline rcgen/src/key_pair.rs:71]: This needs the feature guards from `from_der()` and also another one for `pem`.

File: rcgen/src/key_pair.rs [hunk 6]
```diff
@@ -57,6 +57,90 @@ impl fmt::Debug for KeyPairKind {
...
+			})
+			.ok_or(Error::UnsupportedSignatureAlgorithm)?;
+		Ok(Self {
+			alg,
+			subject_public_key: Vec::from(spki.subject_public_key.as_ref()),
+		})
+	}
+}
+
+#[cfg(feature = "x509-parser")]
```
Comment (djc) [inline rcgen/src/key_pair.rs:108]: This doesn't need the guard.

File: rcgen/src/key_pair.rs [hunk 7]
```diff
@@ -689,7 +773,7 @@ impl<T> ExternalError<T> for Result<T, pem::PemError> {
 	}
 }
 
-pub(crate) trait PublicKeyData {
+pub trait PublicKeyData {
```
Comment (djc) [inline rcgen/src/key_pair.rs:776]: Needs some tweaks, as discussed here, please put those in a separate commit.

https://github.com/rustls/rcgen/issues/282#issuecomment-2206239673

Also needs documentation.

File: rcgen/src/error.rs [hunk 8]
```diff
@@ -45,6 +45,9 @@ pub enum Error {
 	#[cfg(not(feature = "crypto"))]
 	/// Missing serial number
 	MissingSerialNumber,
+	/// X509 parsing error
+	#[cfg(feature = "x509-parser")]
+	X509,
```
Comment (djc) [inline rcgen/src/error.rs:50]: I think we might want to include a `String` representation of the nested error here.

**Review by djc — COMMENTED** [review]
Initial round of feedback.

**Review by est31 — APPROVED** [review]
LGTM if the review comments are addressed

**Review by cpu — COMMENTED** [review]
djc's feedback covered most bases, I just had a couple nits to consider.

**Comment by cpu** [comment]
@kwantam Are you interested in completing this pull-request by addressing the feedback received?

**Comment by djc** [comment]
I've taken this branch and addressed the comments so far in #291.

**Comment by cpu** [comment]
Closing in favour of #291. Thanks for getting the ball rolling Kwantam

**Comment by kwantam** [comment]
Thanks for the comments and @djc for getting this across the finish line! Apologies that I've been buried in other stuff...
